src/librustc_typeck/check/writeback.rs

```diff
@@ -159,8 +160,56 @@ impl<'cx, 'gcx, 'tcx> WritebackCx<'cx, 'gcx, 'tcx> {
             _ => {}
         }
     }
+
+    // Similar to operators, indexing is always assumed to be overloaded
+    // Here, correct cases where an indexing expression can be simplified
+    // to use builtin indexing because the index type is known to be
+    // usize-ish
+    fn fix_index_builtin_expr(&mut self, e: &hir::Expr) {
+        if let hir::ExprIndex(ref base, ref index) = e.node {
+            let mut tables = self.fcx.tables.borrow_mut();
+
+            let base_ty = {
+                let base_ty = tables.expr_ty_adjusted(&base);
+                // When unsizing, the final type of the expression is taken
+                // from the first argument of the indexing operator, which
+                // is a &self, and has to be deconstructed
+                if let TypeVariants::TyRef(_, ref ref_to) = base_ty.sty {
+                    ref_to.ty
+                } else {
+                    base_ty
+                }
+            };
+
+            let index_ty = tables.expr_ty_adjusted(&index);
+            let index_ty = self.fcx.resolve_type_vars_if_possible(&index_ty);
+
+            if base_ty.builtin_index().is_some() && index_ty.is_uint() {
+                // Remove the method call record, which blocks use in
+                // constant or static cases
```

Comment:
This comment is far too specific IMO, without this logic `Index<usize> for [T]` `impl` would infinitely recurse, for example.